{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1P43VCI2OUXTw8qdnUqbCe4xcPwAJGgH3","timestamp":1728416766278}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["*   **Año:** [2024]\n","*   **Alumno/a:** [Manuel Guerrero]\n","*   **Legajo:** [46904766]"],"metadata":{"id":"UwxQU5JCF33j"}},{"cell_type":"markdown","source":["# Pydantic\n","Pydantic es una libreria rapida y extensible que nos permite validar datos usando tipos de datos de Python.\n"],"metadata":{"id":"8ec8DQdbWQfC"}},{"cell_type":"markdown","source":["Primero, importe `pydantic`."],"metadata":{"id":"uF01w4AQcrtV"}},{"cell_type":"code","source":["from pydantic import BaseModel"],"metadata":{"id":"4WPUcwU_c5n2","executionInfo":{"status":"ok","timestamp":1732234032106,"user_tz":180,"elapsed":728,"user":{"displayName":"mg","userId":"10693957844936571260"}}},"execution_count":1,"outputs":[]},{"cell_type":"markdown","source":["Supongamos que tenemos una lista de clientes (llamemoslo en el codigo `Client`). Los clientes tienen dos campos: DNI (un entero) y nacionalidad (un `string`). Cree el modelo base de `Client` (en forma de clases de Python)."],"metadata":{"id":"eC8fpbQ04GzX"}},{"cell_type":"code","source":["class Client(BaseModel):\n","  dni: int\n","  nacionalidad: str"],"metadata":{"id":"I4zJ7s_k4GkI","executionInfo":{"status":"ok","timestamp":1732234034237,"user_tz":180,"elapsed":719,"user":{"displayName":"mg","userId":"10693957844936571260"}}},"execution_count":2,"outputs":[]},{"cell_type":"markdown","source":["Cree a un usuario con documento 39.755.010 y nacionalidad 'Argentina'. Muestre todos sus campos."],"metadata":{"id":"zas-XX4r4SeP"}},{"cell_type":"code","source":["user = Client(dni = 39755010, nacionalidad = \"Argentina\")\n","\n","print(user)"],"metadata":{"id":"_DhtCWhr4ST9","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234322968,"user_tz":180,"elapsed":265,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"8d3fb2b2-bd42-4f4f-d689-7aecab7bdb21"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["dni=39755010 nacionalidad='Argentina'\n"]}]},{"cell_type":"markdown","source":["Intente crear al usuario con un documento en forma de `string`. Deberia fallar..."],"metadata":{"id":"ejbn5RSH4S_W"}},{"cell_type":"code","source":["user = Client(dni = \"39755010\")\n","user = Client(nacionalidad = \"Argentina\")\n","\n","print(user)"],"metadata":{"id":"OMPq8C1C4Su3","colab":{"base_uri":"https://localhost:8080/","height":367},"executionInfo":{"status":"error","timestamp":1732234330668,"user_tz":180,"elapsed":308,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"b5a9cfdb-2b1f-483f-c9a9-2a1d655577d9"},"execution_count":12,"outputs":[{"output_type":"error","ename":"ValidationError","evalue":"1 validation error for Client\nnacionalidad\n  Field required [type=missing, input_value={'dni': '39755010'}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.9/v/missing","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValidationError\u001b[0m                           Traceback (most recent call last)","\u001b[0;32m<ipython-input-12-a24da95ef150>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0muser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdni\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"39755010\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0muser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnacionalidad\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"Argentina\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muser\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pydantic/main.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, **data)\u001b[0m\n\u001b[1;32m    210\u001b[0m         \u001b[0;31m# `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    211\u001b[0m         \u001b[0m__tracebackhide__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 212\u001b[0;31m         \u001b[0mvalidated_self\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__pydantic_validator__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalidate_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself_instance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    213\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mvalidated_self\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    214\u001b[0m             warnings.warn(\n","\u001b[0;31mValidationError\u001b[0m: 1 validation error for Client\nnacionalidad\n  Field required [type=missing, input_value={'dni': '39755010'}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.9/v/missing"]}]},{"cell_type":"markdown","source":["Hemos detectado que ciertos clientes tienen nacionalidades que no existen. Ademas, hay numeros de documento negativos y se tiene que poder agregar la fecha de registro de los clientes (que no pueden ser del futuro). Cambiar la definicion del cliente para que estas cosas no ocurran. Despues de la siguiente celda, cree otras 3 mas probando un caso donde deberia funcionar y otros dos en los que no."],"metadata":{"id":"jARZRMfe6MPE"}},{"cell_type":"code","source":["from pydantic import BaseModel, validator\n","from datetime import date\n","\n","class Client(BaseModel):\n","    dni: int\n","    nacionalidad: str\n","    fecha_registro: date\n","\n","    @validator(\"dni\", pre=True)\n","    def dni_positivo(cls, dni):\n","      if dni < 0:\n","        raise ValueError(\"El DNI debe ser un número positivo\")\n","      return dni\n","\n","    @validator(\"nacionalidad\", pre=True)\n","    def nacionalidad_valida(cls, pais):\n","      nacionalidades_validas = [\"Argentina\", \"Brasil\", \"Chile\", \"Uruguay\", \"Colombia\", \"Peru\",  \"España\", \"Mexico\", \"Estados Unidos\", \"Canada\"]\n","      if pais not in nacionalidades_validas:\n","        raise ValueError(\"Nacionalidad inválida.\")\n","      return pais\n","\n","    @validator(\"fecha_registro\", pre=True)\n","    def fecha_valida(cls, fecha):\n","      if fecha > date.today():\n","        raise ValueError(\"La fecha de registro no puede ser en el futuro.\")\n","      return fecha"],"metadata":{"id":"JWsWZZPH6NFv","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234339094,"user_tz":180,"elapsed":290,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"7524dad9-d127-4540-e824-70d8d4cb2197"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stderr","text":["<ipython-input-13-501bfbc88d34>:9: PydanticDeprecatedSince20: Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style `@field_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/\n","  @validator(\"dni\", pre=True)\n","<ipython-input-13-501bfbc88d34>:15: PydanticDeprecatedSince20: Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style `@field_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/\n","  @validator(\"nacionalidad\", pre=True)\n","<ipython-input-13-501bfbc88d34>:22: PydanticDeprecatedSince20: Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style `@field_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.9/migration/\n","  @validator(\"fecha_registro\", pre=True)\n"]}]},{"cell_type":"code","source":["try:\n","    cliente_valido = Client(dni = 39755010, nacionalidad = \"Argentina\", fecha_registro = date(2023, 1, 1))\n","    print(\"Cliente válido:\", cliente_valido)  # Imprime solo si no hay error\n","except ValueError as exc:\n","    print(\"Error:\", exc)"],"metadata":{"id":"sWyLaqoC7QUK","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234344046,"user_tz":180,"elapsed":282,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"b715edf8-116a-4b70-bba1-33962aa24ab0"},"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["Cliente válido: dni=39755010 nacionalidad='Argentina' fecha_registro=datetime.date(2023, 1, 1)\n"]}]},{"cell_type":"code","source":["try:\n","    cliente_invalido = Client(dni = -123, nacionalidad = \"Brasil\", fecha_registro = date(2022, 5, 10))\n","except ValueError as exc:\n","    print(\"Error:\", exc)"],"metadata":{"id":"E6h_7zb77QhN","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234345591,"user_tz":180,"elapsed":272,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"cf1b307b-14ae-4183-ba4b-f4ed9dc09392"},"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: 1 validation error for Client\n","dni\n","  Value error, El DNI debe ser un número positivo [type=value_error, input_value=-123, input_type=int]\n","    For further information visit https://errors.pydantic.dev/2.9/v/value_error\n"]}]},{"cell_type":"code","source":["try:\n","    cliente_invalido = Client(dni = 45678901, nacionalidad = \"Japon\", fecha_registro = date(2021, 12, 31))\n","except ValueError as exc:\n","    print(\"Error:\", exc)"],"metadata":{"id":"MtDGMxri7R_O","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234347533,"user_tz":180,"elapsed":314,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"6de7d8ea-7215-48a2-de31-792853a30e9e"},"execution_count":16,"outputs":[{"output_type":"stream","name":"stdout","text":["Error: 1 validation error for Client\n","nacionalidad\n","  Value error, Nacionalidad inválida. [type=value_error, input_value='Japon', input_type=str]\n","    For further information visit https://errors.pydantic.dev/2.9/v/value_error\n"]}]},{"cell_type":"markdown","source":["# Pandera\n","Hacer lo que se pidio en la ultima celda anterior, pero con Pandera."],"metadata":{"id":"79P3MNuZ8-t3"}},{"cell_type":"code","source":["!pip install pandera"],"metadata":{"id":"z3DJDdci8-BP","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1732234355223,"user_tz":180,"elapsed":4607,"user":{"displayName":"mg","userId":"10693957844936571260"}},"outputId":"71e24fa1-5cf5-4006-dcbb-9552ec3d1b64"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting pandera\n","  Downloading pandera-0.21.0-py3-none-any.whl.metadata (15 kB)\n","Collecting multimethod (from pandera)\n","  Downloading multimethod-1.12-py3-none-any.whl.metadata (9.6 kB)\n","Requirement already satisfied: numpy>=1.19.0 in /usr/local/lib/python3.10/dist-packages (from pandera) (1.26.4)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pandera) (24.2)\n","Requirement already satisfied: pandas>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from pandera) (2.2.2)\n","Requirement already satisfied: pydantic in /usr/local/lib/python3.10/dist-packages (from pandera) (2.9.2)\n","Requirement already satisfied: typeguard in /usr/local/lib/python3.10/dist-packages (from pandera) (4.4.1)\n","Collecting typing-inspect>=0.6.0 (from pandera)\n","  Downloading typing_inspect-0.9.0-py3-none-any.whl.metadata (1.5 kB)\n","Requirement already satisfied: wrapt in /usr/local/lib/python3.10/dist-packages (from pandera) (1.16.0)\n","Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.2.0->pandera) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.2.0->pandera) (2024.2)\n","Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.2.0->pandera) (2024.2)\n","Collecting mypy-extensions>=0.3.0 (from typing-inspect>=0.6.0->pandera)\n","  Downloading mypy_extensions-1.0.0-py3-none-any.whl.metadata (1.1 kB)\n","Requirement already satisfied: typing-extensions>=3.7.4 in /usr/local/lib/python3.10/dist-packages (from typing-inspect>=0.6.0->pandera) (4.12.2)\n","Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.10/dist-packages (from pydantic->pandera) (0.7.0)\n","Requirement already satisfied: pydantic-core==2.23.4 in /usr/local/lib/python3.10/dist-packages (from pydantic->pandera) (2.23.4)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas>=1.2.0->pandera) (1.16.0)\n","Downloading pandera-0.21.0-py3-none-any.whl (261 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m261.0/261.0 kB\u001b[0m \u001b[31m4.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading typing_inspect-0.9.0-py3-none-any.whl (8.8 kB)\n","Downloading multimethod-1.12-py3-none-any.whl (10 kB)\n","Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)\n","Installing collected packages: mypy-extensions, multimethod, typing-inspect, pandera\n","Successfully installed multimethod-1.12 mypy-extensions-1.0.0 pandera-0.21.0 typing-inspect-0.9.0\n"]}]}]}